src/Host/Broker/Impl/Sessions/SessionManager.cs

```diff
@@ -60,13 +60,17 @@ public class SessionManager {
                 }
 
                 session = new Session(this, user, id, interpreter, commandLineArguments);
-                userSessions.Add(session);
             }
 
             session.StartHost(
                 password,
                 _loggingOptions.LogHostOutput ? _hostOutputLogger : null,
                 _loggingOptions.LogPackets ? _messageLogger : null);
+
+            lock (_sessions) {
+                userSessions.Add(session);
```

Comment:
Good point. I haven't considered this. I think I'll revert it, and we'll just say that if you expect the session to be running, you need to wait until CreateSession completes. For broker clients, this means that you wait until PUT request returns a response before doing GET for a pipe. If you do a GET before PUT completes, the error handling code I've added there should still give reasonable error reporting in that case.

(This change was due to an issue Karthik observed, but it had a different root cause; I just figured I'd patch this one up as well.)

---
In reply to: [75219393](https://github.com/Microsoft/RTVS/pull/2178#discussion_r75219393) [](ancestors = 75219393)